lib/debug/chrome.dart

```diff
@@ -0,0 +1,790 @@
+  }
+}
+
+class ChromeDebugConnection extends ProxyHolder {
+  ChromePage _page;
+  ChromeNetwork _network;
+  ChromeDebugger _debugger;
+  ChromeRuntime _runtime;
+
+  ChromeDebugConnection(JsObject obj) : super(obj);
+
+  ChromePage get page => _page ??= new ChromePage(obj['Page']);
+  ChromeNetwork get network => _network ??= new ChromeNetwork(obj['Network']);
+  ChromeDebugger get debugger => _debugger ??= new ChromeDebugger(obj['Debugger']);
+  ChromeRuntime get runtime => _runtime ??= new ChromeRuntime(obj['Runtime']);
+
+
+  void close() => invoke('close');
+}
+
+class ChromeNetwork extends ProxyHolder {
+
+  ChromeNetwork(JsObject obj) : super(obj);
+
+  Future enable() => promiseToFuture(invoke('enable'));
+  Future disable() => promiseToFuture(invoke('disable'));
+
+  void requestWillBeSent(callback) =>
+      invoke('requestWillBeSent', callback);
+}
+
+class ChromePage extends ProxyHolder {
+
+  ChromePage(JsObject obj) : super(obj);
+
+  Future enable() => promiseToFuture(invoke('enable'));
+  Future disable() => promiseToFuture(invoke('disable'));
+
+  Future reload({bool ignoreCache, String scriptToEvaluateOnLoad}) =>
+      promiseToFuture(invoke('reload', _c({
+        'ignoreCache': ignoreCache,
+        'scriptToEvaluateOnLoad': scriptToEvaluateOnLoad})));
+  Future<String> navigate(String url) =>
+      promiseToFuture(invoke('navigate', {
+        'url': url})).then((obj) => obj['frameId']);
+}
+
+class ChromeDebugger extends ProxyHolder {
+
+  ChromeDebugger(JsObject obj) : super(obj);
+
+  Future enable() => promiseToFuture(invoke('enable'));
+  Future disable() => promiseToFuture(invoke('disable'));
+
+  /// Activates / deactivates all breakpoints on the page.
+  Future setBreakpointsActive(bool active) =>
+      promiseToFuture(invoke('setBreakpointsActive', {'active': active}));
+
+  /// Makes page not interrupt on any pauses (breakpoint, exception,
+  /// dom exception etc).
+  Future setSkipAllPauses(bool skip) =>
+      promiseToFuture(invoke('setSkipAllPauses', {'skip': skip}));
+
+  /// Sets JavaScript breakpoint at given location specified either by URL or
+  /// URL regex. Once this command is issued, all existing parsed scripts will
+  /// have breakpoints resolved and returned in locations property.
+  /// Further matching script parsing will result in subsequent
+  /// breakpointResolved events issued. This logical breakpoint will survive
+  /// page reloads.
+  ///
+  /// [lineNumber]: Line number to set breakpoint at.
+  /// [url]: URL of the resources to set breakpoint on.
+  /// [urlRegex]: Regex pattern for the URLs of the resources to set breakpoints
+  ///   on. Either url or urlRegex must be specified.
+  /// [columnNumber]: Offset in the line to set breakpoint at.
+  /// [condition]: Expression to use as a breakpoint condition. When specified,
+  ///   debugger will only stop on the breakpoint if this expression evaluates
+  ///   to true.
+  Future<Breakpoint> setBreakpointByUrl(int lineNumber,
```

Comment:
Just FYI, in a previous dart web debugger I wrote, we used setBreakpointByRegex. It meant we were less reliant on the url prefix, and on the specific path that the user found to run code from a `packages/` subdir.

Reply:
Can you give me an example of how you built the regex?

Reply:
It would be from memory from a few years ago, but I believe that:
- anything from a package (package: or a self-ref to the lib/ for the current project) was something like `.*/packages/foo_package/foo.dart`
- in a web/ dir (or example/, or test/), we made the regex relative to that dir (`.*/my_app.dart` from /web/my_app.dart)